app/src/ui/branches/branch-list.tsx

```diff
@@ -91,6 +92,13 @@ interface IBranchListProps {
    * to prepopulate the new branch name field.
    */
   readonly onCreateNewBranch?: (name: string) => void
+
+  readonly textbox?: TextBox
+
+  readonly onRenderItem?: (
```

Comment:
This may seem weird, but I needed a way to control how an item is rendered and settled on passing a function that the component can delegate to. If this is the wrong approach, I'd love to chat about it.

Reply:
Render props are a fine way craft higher-order components, which sounds like what you're aiming for. This is already something that we need to support with `List` and it's related components, so it's not a foreign pattern.

To take this further, you could make this a required function and extract the logic for the old code path up into `BranchesContainer`.

Reply:
https://github.com/desktop/desktop/pull/4337